aodntools/timeseries_products/gridded_timeseries.py

```diff
@@ -0,0 +1,262 @@
+    :param depths: list of depth of the instruments at the current timestamp
+    :param max_separation: maximum allowable separation between readings
+    :return: list of depth masked
+    """
+    depth_mask = [False] * len(depth_bins)
+    for depth_index in range(len(depths) - 1):
+        if depths[depth_index + 1] - depths[depth_index] < max_separation:
+            depth_min = bisect.bisect_left(depth_bins, depths[depth_index])
+            depth_max = bisect.bisect_right(depth_bins, depths[depth_index + 1])
+            depth_interval_len = depth_max - depth_min
+            depth_mask[depth_min:depth_max] = [True] * depth_interval_len
+    return depth_mask
+
+def sort_attributes(attributes):
+    """
+
+    :param attributes: dictionary of global attributes
+    :return: sorted dictionary
+    """
+    return dict(sorted(attributes.items()))
+
+def sort_depths(depths, values):
+    """
+    Sort the list of depths and values
+    """
+    index =list(range(len(depths)))
+    index.sort(key=depths.__getitem__)
+    depths[:] = [depths[i] for i in index]
+    values[:] = [values[i] for i in index]
+
+    return depths, values
+
+
+
+
+## MAIN FUNCTION
+def grid_variable(file_name, VoI, depth_bins=None, max_separation=50, depth_bins_increment=10,
+                  base_path='./'):
+    """
+    Grid VoI into depth_bins.
+    :param nc: hourly aggregated dataset with VoI, DEPTH and TIME only
+    :param VoI: variable of interest (TEMP or PSAL)
+    :param depth_bins: list of depth where to interpolate. if null list is provided it will be calculated from the data
+    :param max_separation: max separation allowed for instruments
+    :param depth_bins_increment: in case no depth bins provided this is the increment for the calculated bins
+    :param base_path: path where the result file will be written
+    :return: interpolated dataset
+    """
+
+    with xr.open_dataset(file_name) as nc:
+        site_code = nc.site_code
+        ## get global attributes
+        global_attributes = nc.attrs
+
+        ## get lat/lon
+        longitude_mean = nc.LONGITUDE.mean()
+        latitude_mean = nc.LATITUDE.mean()
+
+        ## in case no depth bins provided, create depth bins to the nearest rounded depth
+        ## if provided, adjust to the min-max registered depth
+        if not depth_bins:
+            depth_bins = make_depth_bins(nc=nc, increment=depth_bins_increment)
+        else:
+            depth_bins = [float(depth) for depth in depth_bins]     # in case depth_bins provided through inline arguments
+            depth_bins = adjust_depth_bins(depth_bins, nc.DEPTH.min(), nc.DEPTH.max())
+
+        ## create empty containers
+        time_0 = pd.to_datetime('1950-01-01T00:00:00')
+        time_min = nc.TIME.values.min()
+        depth_bin_len = len(depth_bins)
+        VoI_binned = xr.DataArray(np.full((depth_bin_len, 1), np.nan), coords=[[time_0], depth_bins],
```

Comment:
maybe is of interest to just check the size before allocating - if depth_bin_len is too big you may explode memory usage.

Reply:
True, but we're not expecting more than ~100 target depths. Even 1000 target depth should be manageable.